### PR TITLE
[react] Allow `ElementType` to be specialized for intrinsic elements

### PR DESCRIPTION
**Pull request by kripod**
When dealing with polymorphic components in React by adding an `as` prop, we may [end up relying upon `React.ElementType<Props = any>`](https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#polymorphic-components-eg-with-as-props). However, that could become quite slow, as TypeScript has to evaluate all the built-in `JSX.IntrinsicElements` for prop matches.

It’s rare that one should allow a polymorphic component to be rendered as _any_ HTML/SVG element. For instance, it’s common for an interactive component to be rendered either as a `button` or an `a` element, but otherwise, if one really wants to render e.g. a `span` or a `div`, they may still do that by specifying a custom React component through the `as` prop.

Existing implementations wouldn’t be affected by adding an optional second parameter to `React.ElementType`, while newer polymorphic components may leverage that for type-level performance advantages.

Therefore, a polymorphic button’s props could be defined as follows:

```ts
type Merge<T, U> = Omit<T, keyof U> & U;

interface MyButtonElementProps {
  className?: string;
}

type MyButtonElementType = React.ElementType<
  MyButtonElementProps,
  "button" | "a" // Where optional type narrowing happens
>;

type MyButtonProps<T extends MyButtonElementType = "button"> =
  Merge<React.ComponentPropsWithoutRef<T>, { as?: T }>;
```

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [<<url here>>](https://react-typescript-cheatsheet.netlify.app/docs/advanced/patterns_by_usecase/#polymorphic-components-eg-with-as-props)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react@latest/))
was missing the following properties:
1. unstable_act
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4107f920220d476a113d6def345456c22cc7a16a
</p>

**Comment by eps1lon** [comment]
Thank you for considering new ways to improve type-checking performance!

You say "Therefore, a polymorphic button’s props could be defined as follows:" but then list `React.ComponentType<MyButtonElementProps, 'a' | 'button'>`. However, your proposed change would not support this syntax. 

Could you add a test for the proposed syntax and contrast that to the pattern that you'd use currently?

**Comment by kripod** [comment]
@eps1lon Thanks for your quick review. I’ve made a mistake in the example and corrected it to `React.ElementType<MyButtonElementProps, 'a' | 'button'>` (from `ComponentType<P, T>`, which is nonexistent.

I’ve also just added tests for the new type parameter.

**Comment by eps1lon** [comment]
> However, that could become quite slow, as TypeScript has to evaluate all the built-in JSX.IntrinsicElements for prop matches.

Do you have some benchmarks to compare before and after?

**Comment by kripod** [comment]
Sure, I’ve just created a repo to showcase the results: https://github.com/kripod/react-specialized-element-type-intrinsics-demo

Let me highlight some results from the comparisons that can be run with the `test:before` and `test:after` scripts, respectively.

### Before

```
Identifiers:                 59467
Symbols:                    160977
Types:                       11471
Instantiations:             357575
Memory used:               134316K
Assignability cache size:     3461
Identity cache size:            26
Subtype cache size:            423
Strict subtype cache size:       2
```

### After

```
Identifiers:                59484
Symbols:                    40024
Types:                       2910
Instantiations:             12050
Memory used:               77305K
Assignability cache size:    1375
Identity cache size:           26
Subtype cache size:            28
Strict subtype cache size:      2
```

File: types/react/index.d.ts [hunk 1]
```diff
@@ -37,10 +37,10 @@ declare namespace React {
     // React Elements
     // ----------------------------------------------------------------------
 
-    type ElementType<P = any> =
+    type ElementType<P = any, T extends keyof JSX.IntrinsicElements = keyof JSX.IntrinsicElements> =
```
[the remark this answers is not in the document]
Comment (kripod) [reply types/react/index.d.ts:40]: Sure we can, I just wanted to follow the existing convention of using single-letter types.

In this case, `T` may refer to `type` or `tag`, just as described by the [docs of `React.createElement`](https://react.dev/reference/react/createElement#createelement).

Would `Tag` be an acceptable name? I’m open to suggestions 🙂

**Review by eps1lon — APPROVED** [review]

**Comment by kripod** [comment]
@eps1lon I have no idea why PR runs are failing. Locally, everything appears to be fine, and there have been no CI issues with the single-letter variant 😕

**Comment by eps1lon** [comment]
CI is unrelated and will be fixed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67582

**Comment by kripod** [comment]
@eps1lon As #67582 has been merged, I’m wondering if the CI could be re-run somehow or I should push an empty commit to trigger a new build 🤔

**Comment by kripod** [comment]
@eps1lon I’ve merged the latest changes from the main branch but react-router-[redux appears to be failing](https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/7091732557/job/19301508685#step:10:1831) now 😕

**Comment by eps1lon** [comment]
> @eps1lon I’ve merged the latest changes from the main branch but [react-router-redux appears to be failing](https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/7091732557/job/19301508685#step:10:1831) now 😕

Should be fixed with https://github.com/DefinitelyTyped/DefinitelyTyped/pull/67765

**Review by peterblazejewicz — APPROVED** [review]
This one looks like already reviewed, just stuck in the process due to CI problems. 
LGTM

**Comment by eps1lon** [comment]
Ready to merge

**Pull request merged**